File: Data Structures & Algorithms/subsets/submission-0.cpp
class Solution {
public:
    // void recurse(int i, 
    // vector<int>&nums, 
    // vector<int>&curSet, 
    // vector<int>&subsets);

    vector<vector<int>> res;

    vector<vector<int>> subsets(vector<int>& nums) {
        
        vector<int> curSet; 

        recurse(0, nums, curSet);
        return res;
        
    }

    void recurse(int n, 
    vector<int>&nums, 
    vector<int>&curSet) {

        res.push_back(curSet);
        for (auto i = n; i < nums.size(); i++) {
            curSet.push_back(nums[i]);
            recurse(i + 1, nums, curSet);
            curSet.pop_back();
        }
    }
};

File: Data Structures & Algorithms/subsets-ii/submission-0.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()

        subsets, cur = [], []

        def recurse(i):
            if i >= len(nums):
                subsets.append(cur.copy())
                return

            # include
            cur.append(nums[i])
            recurse(i + 1)
            cur.pop()
    

            # dont include
            while i + 1 < len(nums) and nums[i] == nums[i + 1]:
                i += 1
            recurse(i + 1)

        recurse(0)
        return subsets

